File: .skills/openclaw-skills/skills/sealawyer2026/skill-evolution-system/ssee/adapters/feishu.py
#!/usr/bin/env python3
"""
飞书平台适配器

对接飞书开放平台
"""

import json
import time
import urllib.request
from pathlib import Path
from typing import Dict, Any, List, Optional
from .base import SkillAdapter, AdapterRegistry


class FeishuAdapter(SkillAdapter):
    """
    飞书平台适配器
    
    对接飞书机器人和应用
    """
    
    BASE_URL = "https://open.feishu.cn/open-apis"
    
    def __init__(self, config: Dict = None):
        super().__init__("feishu", config)
        self.app_id = config.get("app_id", "")
        self.app_secret = config.get("app_secret", "")
        self.tenant_access_token: Optional[str] = None
        self.token_expire_time: float = 0
    
    def connect(self) -> bool:
        """连接到飞书开放平台"""
        if not self.app_id or not self.app_secret:
            return False
        
        try:
            self.tenant_access_token = self._get_tenant_access_token()
            if self.tenant_access_token:
                self._connected = True
                return True
        except Exception as e:
            print(f"飞书连接失败: {e}")
        
        return False
    
    def _get_tenant_access_token(self) -> Optional[str]:
        """获取飞书 tenant_access_token"""
        url = f"{self.BASE_URL}/auth/v3/tenant_access_token/internal"
        data = {
            "app_id": self.app_id,
            "app_secret": self.app_secret,
        }
        
        try:
            req = urllib.request.Request(
                url,
                data=json.dumps(data).encode('utf-8'),
                headers={'Content-Type': 'application/json'},
                method='POST'
            )
            
            with urllib.request.urlopen(req, timeout=10) as response:
                result = json.loads(response.read().decode('utf-8'))
                if result.get("code") == 0:
                    self.token_expire_time = time.time() + result.get("expire", 7200)
                    return result.get("tenant_access_token")
                else:
                    print(f"获取token失败: {result.get('msg')}")
        except Exception as e:
            print(f"请求token失败: {e}")
        
        return None
    
    def _ensure_token_valid(self):
        """确保token有效"""
        if time.time() >= self.token_expire_time - 60:
            self.tenant_access_token = self._get_tenant_access_token()
    
    def _api_call(self, endpoint: str, data: Dict = None, method: str = "GET") -> Dict:
        """调用飞书API"""
        self._ensure_token_valid()
        
        url = f"{self.BASE_URL}{endpoint}"
        
        try:
            headers = {
                'Authorization': f'Bearer {self.tenant_access_token}',
                'Content-Type': 'application/json',
            }
            
            if method == "POST" and data:
                req = urllib.request.Request(
                    url,
                    data=json.dumps(data).encode('utf-8'),
                    headers=headers,
                    method='POST'
                )
            else:
                req = urllib.request.Request(url, headers=headers)
            
            with urllib.request.urlopen(req, timeout=10) as response:
                return json.loads(response.read().decode('utf-8'))
        except Exception as e:
            return {"error": str(e)}
    
    def get_skill_list(self) -> List[Dict]:
        """获取飞书技能列表"""
        # 获取应用列表
        result = self._api_call("/application/v3/apps/list")
        
        if result.get("code") == 0:
            apps = result.get("data", {}).get("apps", [])
            return [
                {
                    "id": app.get("app_id"),
                    "name": app.get("name"),
                    "platform": "feishu",
                    "desc": app.get("description"),
                }
                for app in apps
            ]
        
        return []
    
    def get_skill_metadata(self, skill_id: str) -> Dict:
        """获取技能元数据"""
        result = self._api_call(f"/application/v3/apps/{skill_id}")
        
        if result.get("code") == 0:
            app = result.get("data", {})
            return {
                "id": skill_id,
                "platform": "feishu",
                "name": app.get("name"),
                "desc": app.get("description"),
            }
        
        return {
            "id": skill_id,
            "platform": "feishu",
        }
    
    def track_skill_usage(self, skill_id: str, metrics: Dict) -> Dict:
        """追踪技能使用"""
        return {
            "status": "tracked",
            "skill": skill_id,
            "platform": self.platform_name,
            "metrics": metrics,
        }
    
    def update_skill(self, skill_id: str, updates: Dict) -> bool:
        """更新技能"""
        return True


# 注册适配器
AdapterRegistry.register("feishu", FeishuAdapter)

File: .skills/openclaw-skills/skills/lvjunjie-byte/tiktok-shop-automation/src/feishu.js
/**
 * 飞书集成模块
 * 实现订单同步、通知发送等功能
 * 不依赖 TikTok API，可独立使用
 */

import { loadConfig } from './config.js';

/**
 * 飞书 API 基础 URL
 */
const FEISHU_API_BASE = 'https://open.feishu.cn/open-apis';

/**
 * 飞书集成类
 */
export class FeishuIntegration {
  constructor(config = null) {
    this.config = config || loadConfig();
    this.appToken = this.config.feishu.appToken;
    this.tableId = this.config.feishu.tableId;
    this.webhookUrl = this.config.feishu.webhookUrl || this.config.notifications.feishuWebhook;
  }

  /**
   * 发送 HTTP 请求
   */
  async request(endpoint, options = {}) {
    const url = `${FEISHU_API_BASE}${endpoint}`;
    const headers = {
      'Content-Type': 'application/json',
      ...options.headers
    };

    if (options.accessToken) {
      headers['Authorization'] = `Bearer ${options.accessToken}`;
    }

    const response = await fetch(url, {
      method: options.method || 'GET',
      headers,
      body: options.body ? JSON.stringify(options.body) : null
    });

    const result = await response.json();

    if (!response.ok) {
      throw new Error(`飞书 API 请求失败：${result.msg || response.statusText}`);
    }

    return result;
  }

  /**
   * 获取飞书访问令牌（租户级别）
   * 实际使用时需要通过应用凭证获取
   */
  async getTenantAccessToken(appId, appSecret) {
    console.log('🔐 获取飞书访问令牌...');
    
    // 这里使用 Mock 模式，实际使用需要调用飞书 OAuth API
    // const result = await this.request('/auth/v3/tenant_access_token/internal', {
    //   method: 'POST',
    //   body: { app_id: appId, app_secret: appSecret }
    // });
    
    return {
      tenant_access_token: 'mock_feishu_token_' + Date.now(),
      expire: 7200
    };
  }

  /**
   * 同步订单到飞书多维表格
   * @param {Array} orders - 订单列表
   * @param {Object} options - 选项
   */
  async syncOrdersToBitable(orders, options = {}) {
    console.log('📊 同步订单到飞书多维表格...');
    
    if (!this.appToken || !this.tableId) {
      throw new Error('飞书多维表格配置不完整，请先配置 App Token 和 Table ID');
    }

    const appToken = options.appToken || this.appToken;
    const tableId = options.tableId || this.tableId;

    console.log(`  目标多维表格：${appToken}`);
    console.log(`  数据表 ID: ${tableId}`);
    console.log(`  订单数量：${orders.length}`);

    // 准备数据
    const records = orders.map(order => ({
      fields: {
        '订单 ID': order.order_id,
        '订单状态': this._translateOrderStatus(order.status),
        '订单金额': order.amount,
        '下单时间': order.created_at,
        '客户姓名': order.customer?.name || '',
        '客户邮箱': order.customer?.email || '',
        '商品数量': order.items?.reduce((sum, item) => sum + item.quantity, 0) || 0,
        '物流单号': order.tracking_number || '',
        '同步时间': new Date().toISOString()
      }
    }));

    // Mock 模式：模拟 API 调用
    if (this.config.tiktok?.useMock) {
      console.log('✓ [Mock] 订单同步成功');
      console.log(`  已同步 ${records.length} 条记录`);
      
      return {
        success: true,
        synced: records.length,
        records: records.map((r, i) => ({ record_id: `rec_mock_${i}`, fields: r.fields }))
      };
    }

    // 实际 API 调用（分批写入，每批最多 500 条）
    const batchSize = 500;
    const results = [];

    for (let i = 0; i < records.length; i += batchSize) {
      const batch = records.slice(i, i + batchSize);
      
      const result = await this.request(
        `/bitable/v1/apps/${appToken}/tables/${tableId}/records/batch_create`,
        {
          method: 'POST',
          body: { records: batch }
        }
      );

      results.push(result);
    }

    const totalSynced = results.reduce((sum, r) => sum + (r.data?.records?.length || 0), 0);
    console.log(`✓ 订单同步完成，共 ${totalSynced} 条记录`);

    return {
      success: true,
      synced: totalSynced,
      results
    };
  }

  /**
   * 更新飞书多维表格记录
   * @param {Array} records - 记录列表
   */
  async updateBitableRecords(records, options = {}) {
    console.log('📝 更新飞书多维表格记录...');

    const appToken = options.appToken || this.appToken;
    const tableId = options.tableId || this.tableId;

    // Mock 模式
    if (this.config.tiktok?.useMock) {
      console.log('✓ [Mock] 记录更新成功');
      return {
        success: true,
        updated: records.length
      };
    }

    const result = await this.request(
      `/bitable/v1/apps/${appToken}/tables/${tableId}/records/batch_update`,
      {
        method: 'PUT',
        body: { records }
      }
    );

    console.log(`✓ 更新完成，共 ${result.data?.records?.length || 0} 条记录`);

    return result;
  }

  /**
   * 查询飞书多维表格记录
   * @param {Object} options - 查询选项
   */
  async queryBitableRecords(options = {}) {
    console.log('🔍 查询飞书多维表格记录...');

    const appToken = options.appToken || this.appToken;
    const tableId = options.tableId || this.tableId;
    const filter = options.filter || {};

    // Mock 模式
    if (this.config.tiktok?.useMock) {
      console.log('✓ [Mock] 查询成功');
      return {
        success: true,
        records: [
          {
            record_id: 'rec_mock_1',
            fields: {
              '订单 ID': 'ORD001',
              '订单状态': '待发货',
              '订单金额': 29.99
            }
          }
        ],
        total: 1
      };
    }

    const queryParams = new URLSearchParams();
    if (filter.field_name) {
      queryParams.append('field_name', filter.field_name);
    }
    if (filter.value) {
      queryParams.append('value', filter.value);
    }

    const result = await this.request(
      `/bitable/v1/apps/${appToken}/tables/${tableId}/records?${queryParams}`
    );

    return {
      success: true,
      records: result.data?.items || [],
      total: result.data?.total || 0
    };
  }

  /**
   * 发送飞书消息（通过机器人 Webhook）
   * @param {string} message - 消息内容
   * @param {Object} options - 选项
   */
  async sendWebhookMessage(message, options = {}) {
    console.log('📤 发送飞书消息...');

    const webhookUrl = options.webhookUrl || this.webhookUrl;

    if (!webhookUrl) {
      throw new Error('飞书 Webhook URL 未配置');
    }

    const messageType = options.type || 'text';
    let content;

    if (messageType === 'text') {
      content = {
        msg_type: 'text',
        content: {
          text: message
        }
      };
    } else if (messageType === 'post') {
      content = {
        msg_type: 'post',
        content: {
          post: {
            zh_cn: {
              title: options.title || '通知',
              content: [
                [
                  {
                    tag: 'text',
                    text: message
                  }
                ]
              ]
            }
          }
        }
      };
    } else if (messageType === 'interactive') {
      content = {
        msg_type: 'interactive',
        card: options.card
      };
    }

    // Mock 模式
    if (this.config.tiktok?.useMock) {
      console.log('✓ [Mock] 消息发送成功');
      console.log(`  内容：${message.substring(0, 50)}...`);
      
      return {
        success: true,
        messageId: 'mock_msg_' + Date.now()
      };
    }

    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(content)
    });

    const result = await response.json();

    if (result.StatusCode !== 0 && result.code !== 0) {
      throw new Error(`飞书消息发送失败：${result.msg || result.StatusMessage}`);
    }

    console.log('✓ 消息发送成功');

    return {
      success: true,
      messageId: result.data?.message_id || result.message_id
    };
  }

  /**
   * 发送订单通知
   * @param {Object} order - 订单信息
   */
  async sendOrderNotification(order) {
    const message = `📦 新订单通知

订单 ID: ${order.order_id}
订单金额：¥${order.amount}
下单时间：${new Date(order.created_at).toLocaleString('zh-CN')}
客户：${order.customer?.name || '未知'}
状态：${this._translateOrderStatus(order.status)}

商品:
${order.items?.map(item => `  - ${item.product_id} x ${item.quantity}`).join('\n') || '无'}`;

    return await this.sendWebhookMessage(message, {
      type: 'text'
    });
  }

  /**
   * 发送库存预警通知
   * @param {Object} product - 商品信息
   * @param {number} threshold - 预警阈值
   */
  async sendStockAlert(product, threshold) {
    const message = `⚠️ 库存预警

商品：${product.title}
当前库存：${product.stock}
预警阈值：${threshold}

请及时补货！`;

    return await this.sendWebhookMessage(message, {
      type: 'text'
    });
  }

  /**
   * 发送日报通知
   * @param {Object} report - 日报数据
   */
  async sendDailyReport(report) {
    const message = `📊 每日销售日报

日期：${report.date}
总销售额：¥${report.totalSales}
总订单数：${report.totalOrders}
转化率：${report.conversionRate}%

详细报告请查看附件或登录后台查看。`;

    return await this.sendWebhookMessage(message, {
      type: 'text'
    });
  }

  /**
   * 翻译订单状态
   */
  _translateOrderStatus(status) {
    const statusMap = {
      'pending': '待处理',
      'processing': '处理中',
      'shipped': '已发货',
      'delivered': '已送达',
      'cancelled': '已取消',
      'refunded': '已退款'
    };
    return statusMap[status] || status;
  }

  /**
   * 测试飞书连接
   */
  async testConnection() {
    console.log('🔌 测试飞书连接...');

    try {
      // Mock 模式直接返回成功
      if (this.config.tiktok?.useMock) {
        console.log('✓ [Mock] 飞书连接测试成功');
        return { success: true, message: 'Mock 模式' };
      }

      // 实际连接测试
      if (this.webhookUrl) {
        await this.sendWebhookMessage('TikTok Shop Automation 连接测试', {
          type: 'text'
        });
        console.log('✓ 飞书连接测试成功');
        return { success: true, message: '连接正常' };
      } else {
        console.log('⚠️  未配置 Webhook URL');
        return { success: false, message: '未配置 Webhook URL' };
      }
    } catch (error) {
      console.error('✗ 飞书连接测试失败:', error.message);
      return { success: false, message: error.message };
    }
  }
}

/**
 * 创建飞书集成实例
 */
export function createFeishuIntegration(config) {
  return new FeishuIntegration(config);
}

export default FeishuIntegration;
